docs/proposals/29-autoscaling-storage-volumes.md

```diff
@@ -0,0 +1,326 @@
+[[1]], hereafter referred to as `pvc-autoscaler`, is used. It is
+integrated as a seed system component, and runs as a deployment in the seed's `garden` namespace.
+Its primary driving signal is the PVC metrics from the seed's cache Prometheus instance, which it
+periodically examines, and if capacity is found to be near exhaustion, `pvc-autoscaler` takes action by updating the
+PVC's storage request.
+
+In this initial iteration, `pvc-autoscaler` scales two categories of `vali`, `prometheus`, and `alertmanager` volumes:
+those in shoot namespaces, and those in the seed's `garden` namespace.
+
+`pvc-autoscaler` publishes Prometheus metrics to anonymous scrapers via a `/metrics` HTTP endpoint. A `ServiceMonitor`
+object is created in the `garden` namespace, and drives the seed's `prometheus-operator` to configure a scrape on that
+endpoint by `prometheus-seed`.
+
+`pvc-autoscaler` runs in active/passive replication mode, based on the standard leader election mechanism, supplied by
+the K8s controller runtime library. Only the active replica is shown in _Fig.1_.
+
+### PVC Autoscaler - Deployer Structure
+The `pvc-autoscaler` application and its supporting artefacts are deployed by `gardener`, as part of the seed
+reconciliation flow.
+
+The `vali` and `prometheus` deployers, part of `gardenlet`'s shoot reconciliation flow, ensure that the PVC template section in
+the respective StatefulSet objects contains the annotations, necessary to instruct `pvc-autuscaler` how to act on the
+PVCs, created from those templates.
+
+That `gardenlet` action mechanism, based on PVC templates, only affects PVCs upon creation. Therefore, it has no effect
+on existing shoots, and on the observability PVCs of existing seed clusters. To enable PVC autoscaling for existing
+shoots, those PVCs will be directly annotated by the operator. Such direct intervention allows gradual transition,
+where autoscaling can initially be enabled only for a few PVCs for validation purposes.
+Similarly, to enable PVC autoscaling on existing seed clusters' `garden` PVCs, those will be directly annotated at
+operators' discretion.
+
+![02-operator_structure.png](assets/gep29-02-operator_structure.png)
+
+_Fig.2: Operator structure_
+
+VPA is used to scale `pvc-autoscaler`. For the future security enhancement scenario, where `kube-rbac-proxy` is used
+(see [section 'Future Enhancements'](#future-enhancements)),
+that container is expected to be excluded from autoscaling, and to run with fixed resource requests. Pod evictions
+driven by that lean container are unjustified, as they would only achieve minimal resource savings, and disrupt
+the main container.
+
+The feature is deployed behind a dedicated feature gate. When disabled, `pvc-autoscaler` is removed, and related
+annotations are removed from the shoot control plane observability StatefulSets, and the seed observability StatefulSets.
```

Comment:
If I understand correctly, human operators have to annotate PVCs (should better be labels for easier filtering) they want to be auto-scaled. Is this understanding correct? If so, do we need a feature gate at all? Wouldn't removing the label be good enough?

Reply:
As soon as we enable the new feature, each new shoot, created after that point on, comes out in the new shape (observability volumes automatically annotated, small initial size, and PVC autoscaling enabled). Preexisting shoots are not affected, and require annotating manually, but the shoot creation process is immediately affected.

Reply:
Interesting. Why are pre-existing shoots not affected? Who/what mechanism will label the volumes? What's the different with new shoots?

Reply:
>what mechanism
We're creating PVCs via a PVC template in the STS. Not direct reconciliation. So once a PVC is created, the template doesn't affect it. The implementation would update the PVC template with the autoscaler-enabling annotations, and leave the rest to the operator.

I like how that works out in the end, leaving control in the hands of the operator. For one thing, existing PVs are likely heavily underutilised, so enabling autoscaling won't be doing much for them. If the operator wants to save on storage, it's up to them to delete the volumes, so they can start anew, from a small size.

There's actually a piece of existing code in `gardenlet`, which resizes PVCs upon reconciliation, if a new size is specified (as a parameter(value) to the deployer). But I plan to disable it when autoscaling is active. For downsizing, the existing code deletes the PV. For upsizing, it relies on the storage class supporting expansion, which it does not check. So it was just a quick fix.

Reply:
I don't like the approach that a human operator needs to do things manually beyond activating the feature gate. We always strive for automation where possible and reasonable. In this light, what you propose is inconsistent with what happens for newly created shoots and creates additional overhead. As an operator using this feature gate, I would expect all shoots (new and existing) to be considered for PVA. Hence, `gardenlet` should also update existing PVCs with the label, even if it might not have an effect since most of them might be underutilised.

Reply:
During offline discussion, @rfranzke expressed preference that `pvc-autoscaler` be refactored to use a new custom object as source of per-PVC settings, instead of the current annotations in the PVC object itself. The goal would be, improve UX by building a parallel with the existing VPA.

In a discussion with the `pvc-autoscaler` authors, they expressed an open stance toward the idea of a per-PVC configuration object as a possible future enhancement, with the note that `pvc-autoscaler` is not under active development at the moment. Delaying the present GEP to wait for a substantial rework of `pvc-autoscaler` thus seems counterproductive: the `gardenlet` feature proposed by the GEP has no user exposure, and can be refactored at any time in the future, at minimal cost.

As a side benefit from the meeting with the `pvc-autoscaler` authors, it was noted that this GEP should explicitly communicate our vision regarding migration of existing storage volumes to autoscaling. To that end, I added a small new section in the GEP, dedicated to that topic. Currently if follows the GEP's original view of deferring all responsibility to the operator, as it seems to coincide with `pvc-authoscaler`' authors' vision. However, this decision should ultimately be made in the present forum. The new section is thus only meant to serve as a basis for discussion, and is open to change.

Reply:
@rfranzke @ScheererJ @plkokanov 
A consultation with with observability folk (@nickytd, @dnaeon) confirmed that there are currently no plans to provide automatic migration for preexisting PVs.  Obviously we can't just drop the data either. Therefore the best we can do at this time is what you requested above: automatically enable autoscaling on preexisting PVs (and we leave them at their current size).

In the latest commit I updated the GEP to automatically annotate existing `vali` and `prometheus` PVCs.

Reply:
It is great that handling existing PVCs are in scope now. Yet, I am unhappy with the refusal of the discussed enhancement of PVA w.r.t. switching its annotation APIs to CRDs. In an offline discussion, @ScheererJ shared similar thoughts, but perhaps he can also speak up here.

> In a discussion with the `pvc-autoscaler` authors, they expressed an open stance toward the idea of a per-PVC configuration object as a possible future enhancement, with the note that `pvc-autoscaler` is not under active development at the moment. Delaying the present GEP to wait for a substantial rework of `pvc-autoscaler` thus seems counterproductive: the `gardenlet` feature proposed by the GEP has no user exposure, and can be refactored at any time in the future, at minimal cost.

I disagree with the statement that switching from annotations to CRD is "substantial rework" or "counter-productive". It is also no excuse that this component is "not under active development". Such a change can be done with a small number of development hours. This is still our component in the `gardener` org, so we have full control of it and can decide to put it into "active development" for a day or two if you will.

As you seem to agree that this is actually a good thing/a reasonable enhancement, I don't see any satisfying reason to not to this now before PVA gets integrated and rolled out to existing systems. Refactoring this later, even at "minimal cost", is just waste and extra-effort. It can be done already now, effectively saving additional work in the future.

Reply:
>As you seem to agree that this is actually a good thing(...)

The difficulty when joining data from multiple sources is something which caused me to reverse my preference for this case.

In general, I favour modularisation, and data models and REST APIs are no exception. In our case, that modular design would be the one based on a CRD. However, I recently realised that the lack of support for joining data makes 1:1 object relationships quite harmful to UX in the K8s CLI case (kubectl). Splitting objects into custom resources turns one-liners into highly inconvenient for+for+if manual joins. 

It's also worth noting that the analogy with VPA isn't complete, from a UX perspective. VPA is not a small type, and Pod is particularly large. So merging those would make for an overwhelmingly large object. PVC on the other hand is a very small object, and a PVCAutoscaler spec would consist of (IIRC) just 5 primitive values, 4 of which are very optional. Aside from reporting back status, PVCAutoscaler is likely to end up as an entire K8s resource which configures just "Enabled = true". 

A dedicated CRD would allow for a much better extensibility. IMO, a CRD would be particularly useful if we decide to implement multiple scaling strategies/algorithms, each with a different set of parameters. However, I do not know that we would make use of such extensibility.

So, all things considered, if I was faced with the choice whether to base `pvc-autoscaler` on annotations or a CRD, I'd be on the fence. Since we already have implementations based on one of those options, I find it more economical to go with that option. In other words, follow the original plan to deliver a MVP, and only rework/extend if we find it necessary.

Reply:
I am sharing my thoughts on the CRD vs annotations discussion without imposing a blocker for a rework of the existing pvc-autoscaler:

A major pro of the CRD approach is its status. I haven't read the full GEP, neither looked into the implementation of the pvc-autoscaler. But, my assumption is that the pvc-autoscaler creates events on successful/failed resize operations. If that's the case, events are ephemeral and they get deleted when they ttl expires (defaults to 1h for a Gardener Shoot). When a failed resize operation occurs, then the pvc-autoscaler can reflect this in the CR status. Then, other components like gardenlet can health check the custom resource. This is similar to how gardenlet is health checking VPA custom resources and reflecting issues with them in the Shoot status. Otherwise, how can an operator can be noticed about failed resize operations that occur in the landscape?

tl;dr: IMO, a bigger pro argument for the CRD approach is not its spec, but rather it is status. An operator can read the status for investigation/debugging purposes. gardenlet can read the status and then reflect failures to the Shoot status.

Reply:
> @andrerun Sorry, but I do not get your points.
> 
> Annotations are a non-typed interface, which can easily pose problems. Just think of a little typo in the key or value, which is very easy to overlook. Even if there are currently only 5 primitive values to configure that certainly makes sense to be put into a separate CRD so that you can configure it properly with type safety and checks.
> 
> How would you even envision this with annotation? Would you put five annotations on a resource or have a `json` value for an annotation? Both approaches are rather bad from my point of view.
> 
> However, just do not only take my word for it. Feel free to check what Tim Hockin for example says with regards to annotations in the ingress API in the [kubernetes podcast](https://kubernetespodcast.com/episode/041-ingress/).
> 
> Especially as you have put it as the long term vision to also provide volume autoscaling for shoot workloads, I think it is necessary to have a proper API (read: CRD) and not just annotations. It also directly allows you to provide a proper status to your consumers.
> 
> From my point of view, the current approach with annotations is a blocker. Please reconsider.

Hey @ScheererJ ,

I don't see how this is any different from the existing implementation of Network Policies controller, which is managed by a mix of labels + annotations, and on top of that -- embeds JSON within these annotations.

References:

- https://gardener.cloud/docs/gardener/network_policies/
- https://github.com/gardener/gardener/issues/7352

This is an example annotation which magically configures a network policy.

``` yaml
annotations:
  networking.resources.gardener.cloud/from-all-garden-scrape-targets-allowed-ports: '[{"port":<metrics-port-on-pod>,"protocol":"<protocol, typically TCP>"}]'
```

According to what you've mentioned this is a bad design, however we are already promoting such design (and implementation) to Gardener users with the use of the network policies annotations + labels.

I'm not picking on the Network Policies controller here, I'm simply trying to understand the rationale here and the criteria against you are measuring one annotations-approach as good, and another one as bad? @ScheererJ , @rfranzke ?

For the sake of moving forwarding with the integration of pvc-autoscaler into g/g I've also merged CRDs support into `pvc-autoscaler`.

However, please elaborate on the question I've asked above, so that myself (and others included) can better understand the design decisions and rationale which make annotations sometimes good, and other times bad.

Thanks!

Reply:
Hey @dnaeon,

your example is spot on. Some of the network policy controller annotations use structured data (read: JSON) and I am not too happy about it. Unfortunately, it is not always clear cut in good/bad. As usual, it depends.

Why do I find it more acceptable (but still not great) to do it like this for the network policy controller? This API is purely for Gardener internal usage. Yes, extension implementors may need to deal with it, but that is rather a limit amount of people. In addition to that, there is no need to provide any status values.

The network policy controller is meant for simplifying the network policy creation as it is hard to create the correct network policies (as history/experience shows). The abstraction over the network policy API was done with annotations/labels and evolved from simple to more complicated use cases. If I remember correctly, the initial solution did not include the complex JSON values, but @rfranzke may remember better.

If we would consider making the network policy controller available to cluster owners I would probably lean more towards a new CRD instead of exposing our label/annotation API.

TLDR: It always depends. The bigger your potential API audience is the simpler/cleaner/better your API should be. PVC autoscaling has large goals (all Gardener clusters) while network policy controller only wants to rule the Garden/Seeds.

Hope that helps even though it is not clear cut.

Thanks a lot for implementing the CRD support in `pvc-autoscaler`. Let's adapt the GEP accordingly and finish the remaining smaller discussions so that we can move forward.

@andrerun Please check the commit messages. The bot is currently complaining, which would block a merge.

Reply:
To record an answer which was part of an offline discussion:
`pvc-autoscaler` uses a status condition on the PVC to reflect the error/success status of the last resize.

Reply:
Gardener core team effectively resolved this discussion by posting an overruling requirement to use the approach based on a custom resource. For details, see [this comment](https://github.com/gardener/gardener/pull/10690#issuecomment-2535767407).